plugin/plugin.go

```diff
@@ -363,6 +378,296 @@ func RegisterShutdown(name string, s Shutter) error {
 	return nil
 }
 
+// Configuration defines the interface for complex_configure callbacks, i.e. Go
+// structs that can hold and validate a configuration
+type Configuration interface {
+	Validate() error
+}
+
+var config_target Configuration
+var config_targets map[string]Configuration
+
+var configured map[string]chan struct{}
+
+// getOconfigChildren takes a C.oconfig_item_t and returns an array of Go
+// pointers to its child items. Getting the child pointers requires a little
+// more arithmetic than in C as we can't just increment a pointer to get the
+// next array item
+func getOconfigChildren(oconfig *C.oconfig_item_t) (output []*C.oconfig_item_t) {
+	start := unsafe.Pointer(oconfig.children)
+	size := unsafe.Sizeof(*oconfig.children)
+	for i := 0; i < int(oconfig.children_num); i++ {
+		child := (*C.oconfig_item_t)(unsafe.Pointer(uintptr(start) + size*uintptr(i)))
+		output = append(output, child)
+	}
+	return
+}
+
+// checkMatchable returns an error if the provided source cannot be copied to
+// the target, or nil if it's OK
+func checkMatchable(oconfig *C.oconfig_item_t, target reflect.Value) error {
+	if oconfig.children_num > 0 { // if it has children we need a corresponding struct
+		if target.Kind() != reflect.Struct {
+			return fmt.Errorf("did not have corresponding struct")
+		}
+	}
+	if oconfig.values_num > 1 { // if it has multiple values we need a slice
```

Comment:
This section may not take care of case where the PlugIn stanza's root struct has string. For example,

```
<Plugin MyCoolPlugin>
  <MyRoot "myString">
     Config1 "String"
     Config2 123
  </MyRoot>
</Plugin>
```